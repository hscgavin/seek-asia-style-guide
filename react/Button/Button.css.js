import {buttonPrimary, buttonDefault, buttonSuccess} from '../../theme/palette/palette.js';

export default {
	'.primary': {
		backgroundColor: buttonPrimary
	},
	'.default': {
		backgroundColor: buttonDefault
	},
	'.success': {
		backgroundColor: buttonSuccess
	}
}