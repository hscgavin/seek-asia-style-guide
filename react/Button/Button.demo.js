import React from 'react';
import { Button, HeartIcon, StarIcon } from 'seek-style-guide/react';
import styles from './Button.less';
import classnames from 'classnames';

export default {
  route: '/button',
  title: 'Button',
  component: Button,
  initialProps: {
    children: 'Hello world',
    color: 'primary'
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Hover',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootHover)
          })
        },
        {
          label: 'Active',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootActive)
          })
        },
        {
          label: 'Focus',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootFocus)
          })
        },
        {
          label: 'Loading',
          transformProps: props => ({
            ...props,
            loading: true
          })
        },
        {
          label: 'Full width',
          transformProps: props => ({
            ...props,
            fullWidth: true
          })
        }
      ]
    },
    {
      label: 'Component',
      type: 'radio',
      states: [
        {
          label: 'As Button',
          transformProps: props => ({
            ...props,
            component: 'button'
          })
        },
        {
          label: 'As Link',
          transformProps: props => ({
            ...props,
            component: 'a',
            href: '#'
          })
        },
        {
          label: 'As Div',
          transformProps: props => ({
            ...props,
            component: 'div'
          })
        }
      ]
    },
    {
      label: 'Colour',
      type: 'radio',
      states: [
        {
          label: 'Primary',
          transformProps: props => ({
            ...props,
            color: 'primary'
          })
        },
        {
          label: 'Default',
          transformProps: props => ({
            ...props,
            color: 'default'
          })
        },
        {
          label: 'Success',
          transformProps: props => ({
            ...props,
            color: 'success'
          })
        },
        {
          label: 'Transparent',
          transformProps: props => ({
            ...props,
            color: 'transparent'
          })
        }
      ]
    },
    {
      label: 'Icon',
      type: 'radio',
      states: [
        {
          label: 'No Icon',
          transformProps: props => props
        },
        {
          label: 'Heart Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <HeartIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        },
        {
          label: 'Star Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <StarIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        }
      ]
    }
  ]
};
