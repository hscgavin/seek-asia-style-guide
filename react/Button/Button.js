import styles from './Button.less';
import brandStyles from './Button.css.js';

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import classnames from 'classnames';

export default class Button extends Component {

  static displayName = 'Button';

  static propTypes = {
    color: PropTypes.oneOf([
      'callToAction', 'hyperlink', 'completion', 'alert', 'highlight', 'transparent'
    ]).isRequired,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired,
    className: PropTypes.string,
    component: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.string
    ]),
    loading: PropTypes.bool,
    fullWidth: PropTypes.bool
  };

  static defaultProps = {
    className: '',
    loading: false,
    fullWidth: false,
    component: 'button'
  };

  constructor() {
    super();

    this.storeButtonReference = this.storeButtonReference.bind(this);
  }

  storeButtonReference(button) {
    if (button !== null) {
      this.button = button;
    }
  }

  render() {
    const { color, className, loading, fullWidth, children, component, ...restProps } = this.props;

    const combinedProps = {
      className: classnames(styles.root, className, {
        [styles.loading]: loading,
        [styles.fullWidth]: fullWidth,
        [brandStyles.callToAction]: color === 'callToAction',
        [brandStyles.hyperlink]: color === 'hyperlink',
        [brandStyles.completion]: color === 'completion',
        [brandStyles.alert]: color === 'alert',
        [brandStyles.highlight]: color === 'highlight',
        [styles.root_isTransparent]: color === 'transparent'
      }),
      disabled: loading,
      ref: this.storeButtonReference,
      ...restProps
    };


    return React.createElement(component, combinedProps, children);
  }

}
