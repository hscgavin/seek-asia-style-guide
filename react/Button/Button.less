@import (reference) "~seek-asia-style-guide/theme";

@keyframes loading {
  to { width: @grid-gutter-width; }
}

.hoverState(@rules) {
  &:hover,
  &.rootHover {
    @rules();
  }
}

.activeState(@rules) {
  &:active,
  &.rootActive {
    @rules();
  }
}

.focusState(@rules) {
  &:focus,
  &.rootFocus {
    @rules();
  }
}

.buttonColor(@color, @background-color) {
  color: @color;
  background-color: @background-color;
  .hoverState({
    background-color: darken(@background-color, 5%);
  });  
}

.root {
  .touchableText();
  display: inline-block;
  text-decoration: none;
  user-select: none;
  text-align: center;
  padding: 0 @grid-gutter-width;  
  border: 0;
  border-radius: 3px;
  transition:
    background-color 0.2s,
    transform 0.2s cubic-bezier(0.02, 1.505, 0.745, 1.235);
  cursor: pointer;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;  
  font-weight: bold;  
  .activeState({
    outline: none;
  });
  .focusState({
    .focus();
    box-shadow:none;
  });
  @media @desktop {
    .touchableText(@standard-type-scale);
    line-height: 38px;
    height: 38px;
  }
}

.root_callToAction{
  .buttonColor(@sk-white, @saBlue);
}
.jobStreet.root_callToAction{
  .buttonColor(@sk-black, @actionYellow);  
}
.jobsDB.root_callToAction{
  .buttonColor(@sk-white, @actionOrange);
}
.root_hyperlink{
  .buttonColor(@sk-white, @actionBlue);
}
.root_completion{
  .buttonColor(@sk-white, @actionGreen);
}
.root_alert{
  .buttonColor(@sk-white, @actionRed);
}
.root_highlight{
  .buttonColor(@saGrey2, @actionPastelBlue);
}

.root_isPink {
  .buttonColor(@sk-white, @sk-pink);
}

.root_isBlue {
  .buttonColor(@sk-white, @sk-highlight);
}

.root_isGray {
  .buttonColor(@sk-black, @sk-mid-gray-light);
}

.root_isTransparent {
  color: @sk-link;
  background-color: transparent;
  box-shadow: none;
  padding-left: 0;
  padding-right: 0;
  .hoverState({
    text-decoration: underline;
  });
  .activeState({
    transform: none;
  });
  @media @desktop {
    .touchableText(@standard-type-scale);
  }
}

.loading {
  padding-right: @grid-gutter-width * 2;

  &::after {
    content: ' \2026';
    text-align: left;
    width: 0;
    overflow-x: hidden;
    display: inline-block;
    position: absolute;
    animation-duration: 0.8s;
    animation-name: loading;
    animation-iteration-count: infinite;
  }
}

.fullWidth {
  width: 100%;
}

.group {
  width: 100%;

  .root {
    vertical-align: top;
    @media only screen and (max-width: 500px) {
      width: 100%;
    }
    &:not(.root_isTransparent) {
      @media @mobile {
        padding-left: @grid-gutter-width * 2;
        padding-right: @grid-gutter-width * 2;
      }
    }
    & + .root {
      margin-left: @grid-gutter-width / 2;
      @media only screen and (max-width: 500px) {
        margin-top: @grid-row-height;
        margin-left: 0;
      }
    }
    & + .root_isTransparent {
      margin-left: @grid-gutter-width * 2;
      @media only screen and (max-width: 500px) {
        margin-left: 0;
      }
    }
  }
}
