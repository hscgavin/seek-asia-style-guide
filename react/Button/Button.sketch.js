import React from 'react';
import Button from './Button';

export const symbols = {
  'Button/Call To Action': <Button color="callToAction">Button</Button>,
  'Button/Hyperlink': <Button color="hyperlink">Button</Button>,
  'Button/Completion': <Button color="completion">Button</Button>,
  'Button/Alert': <Button color="alert">Button</Button>,
  'Button/Highlight': <Button color="highlight">Highlight</Button>,
  'Button/Transparent': <Button color="transparent">Button</Button>
};
