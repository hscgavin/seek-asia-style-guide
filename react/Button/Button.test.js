import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button:', () => {
  describe('color variants:', () => {
    it('should render primary button', () => {
      const wrapper = shallow(<Button color="primary">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render default button', () => {
      const wrapper = shallow(<Button color="default">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render success button', () => {
      const wrapper = shallow(<Button color="success">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render transparent button', () => {
      const wrapper = shallow(<Button color="transparent">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should render with node', () => {
    const wrapper = shallow(<Button color="primary"><h5>SEEK</h5></Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with array of nodes', () => {
    const wrapper = shallow(<Button color="primary"><span>SEEK</span><span>AU</span></Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as an anchor with href="https://www.seek.com.au"', () => {
    const wrapper = shallow(<Button color="default" component="a" href="https://www.seek.com.au">SEEK</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with className', () => {
    const wrapper = shallow(<Button color="success" className="foo">SEEK</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render fullwidth button', () => {
    const wrapper = shallow(<Button color="transparent" fullWidth>SEEK</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as loading', () => {
    const wrapper = shallow(<Button color="primary" loading>SEEK</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render custom component based on itâ€™s reference', () => {
    const CustomComponent = props => <a {...props} someProp="kraken" />;
    const wrapper = shallow(<Button color="default" component={CustomComponent}>SEEK</Button>);
    expect(wrapper).toMatchSnapshot();
  });
});
