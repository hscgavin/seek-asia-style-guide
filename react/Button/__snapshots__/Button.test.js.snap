// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: color variants: should render alert button 1`] = `
<button
  className="root root_alert"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render callToAction button 1`] = `
<button
  className="root root_callToAction"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render completion button 1`] = `
<button
  className="root root_completion"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render highlight button 1`] = `
<button
  className="root root_highlight"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render hyperlink color button 1`] = `
<button
  className="root root_hyperlink"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render transparent button 1`] = `
<button
  className="root root_isTransparent"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render white ghost button 1`] = `
<button
  className="root root_isGhost root_isWhite"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render as an anchor with href="https://www.seek.com.au" 1`] = `
<a
  className="root root_hyperlink"
  disabled={false}
  href="https://www.seek.com.au"
>
  SEEK
</a>
`;

exports[`Button: should render as loading 1`] = `
<button
  className="root loading root_callToAction"
  disabled={true}
>
  SEEK
</button>
`;

exports[`Button: should render custom component based on itâ€™s reference 1`] = `
<CustomComponent
  className="root root_callToAction"
  disabled={false}
>
  SEEK
</CustomComponent>
`;

exports[`Button: should render fullwidth button 1`] = `
<button
  className="root fullWidth root_isTransparent"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render with array of nodes 1`] = `
<button
  className="root root_callToAction"
  disabled={false}
>
  <span>
    SEEK
  </span>
  <span>
    AU
  </span>
</button>
`;

exports[`Button: should render with className 1`] = `
<button
  className="root foo root_callToAction"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render with node 1`] = `
<button
  className="root root_callToAction"
  disabled={false}
>
  <h5>
    SEEK
  </h5>
</button>
`;
